### DISCUSSION TITLE
Testclient does not treat readable data correctly

### DISCUSSION THREAD

**Discussion by plankthom** [encode/starlette#1561] [Potential Issue] [answered]
I think this is a potential issue with the TestClient 

When posting streaming data, a client call like
```python
with open(archive_path, 'rb') as archive_file:
    requests.put(url, headers={'Content-Type': 'application/gzip'}, data=archive_file) 
```
can be processed by a server route with a helper method
```python
async def _write_request_stream(request: Request, target: BufferedWriter):
     async for data in request.stream():
         target.write(data)
```

However this doesn't work with the TestClient, i needed the following workaround on the server
```python
from shutil import copyfileobj as copybuffers

async def _write_request_stream(request: Request, target: BufferedWriter):
     async for data in request.stream():
        if isinstance(data, BufferedReader):
            copybuffers(data, target)
        else:
            target.write(data)
```

IMO the test client should treat body arguments with a `read` method as the `requests` library does, converting it to chunks:
https://github.com/python/cpython/blob/c26af2bc531eb114c378cdad81935f6e838a7ee0/Lib/http/client.py#L1045 

There seems to be some handling for a request body that is a generator, See https://github.com/encode/starlette/blob/87641dc8e7e72755f1b8f39dfc9b3bfbdf41c923/starlette/testclient.py#L212 ,
but IMO this should be extended to all `body` arguments that have a `read` method (for compatibility with the `requests` library)

Shall I create an issue , and maybe contribute a PR?

**Comment by Kludex** [comment]
We changed the HTTP client from requests to `httpx`, is this still an issue?

**Comment by Kludex** [comment, marked as answer]
I'll preemptively close this as outdated, but please let me know.
